Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,321 @@
+            // _then_ shrink the rest of the buffer onto it.
+            self._moveBytes(from: originalMessageEndIndex, to: lengthPrefixIndex, size: actualIntegerLength)
+            let newMessageStartIndex = lengthPrefixIndex + actualIntegerLength
+            self._moveBytes(
+                from: originalMessageStartIndex,
+                to: newMessageStartIndex,
+                size: messageLength
+            )
+            self.moveWriterIndex(to: newMessageStartIndex + messageLength)
+        case reservedCapacity...:
+            // We wrote more bytes. We now have to create enough space. Once we do, we have the same
+            // implementation as the matching case.
+            let extraSpaceNeeded = actualIntegerLength - reservedCapacity
+            self._createSpace(before: lengthPrefixIndex, requiredSpace: extraSpaceNeeded)
+
+            // Clean up the indices.
+            let newMessageEndIndex = originalMessageEndIndex + extraSpaceNeeded
+            // We wrote the length after the message, so we have to move those bytes to the space at the front
+            self._moveBytes(from: newMessageEndIndex, to: lengthPrefixIndex, size: actualIntegerLength)
+            self.moveWriterIndex(to: newMessageEndIndex)
+        default:
+            fatalError("Unreachable")
+        }
+
+        let totalBytesWritten = self.writerIndex - lengthPrefixIndex
+        return totalBytesWritten
+    }
+
+    /// Reads a slice which is prefixed with a length. The length will be read using `strategy`, and then that many bytes will be read to create a slice.
+    /// - Returns: The slice, if there are enough bytes to read it fully. In this case, the readerIndex will move to after the slice.
+    /// If there are not enough bytes to read the full slice, the readerIndex will stay unchanged.
+    public mutating func readLengthPrefixedSlice<Strategy: NIOBinaryIntegerEncodingStrategy>(
+        _ strategy: Strategy
+    ) -> ByteBuffer? {
+        let originalReaderIndex = self.readerIndex
+        guard let length = strategy.readInteger(as: Int.self, from: &self), let slice = self.readSlice(length: length)
+        else {
+            self.moveReaderIndex(to: originalReaderIndex)
+            return nil
+        }
+        return slice
+    }
+}
+
+// MARK: - Helpers for writing length-prefixed things
+
+extension ByteBuffer {
+    /// Write the length of `buffer` using `strategy`. Then write the buffer.
+    /// - Parameters:
+    ///   - buffer: The buffer to be written.
+    ///   - strategy: The encoding strategy to use.
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the buffer itself.
+    @discardableResult
+    public mutating func writeLengthPrefixedBuffer<
+        Strategy: NIOBinaryIntegerEncodingStrategy
+    >(
+        _ buffer: ByteBuffer,
+        strategy: Strategy
+    ) -> Int {
+        var written = 0
```

Comment:
It might be worth reserving enough capacity ahead of time here (and for the `String` version too)

Reply:
we don't know how many bytes writeEncodedInteger will need

We could make use of strategy.reservedCapacity here, but it might be wrong

Reply:
Being wrong by a couple of bytes should be fine because we round up to the next power of two anyway. Reserving capacity would help us to avoid situations where we allocate a couple of bytes for the length then immediately reallocate because we need more space to write the buffer after it.

Reply:
ok, done